### ISSUE TITLE
hasFieldOrPropertyWithValue() eats exceptions thrown by getters, and reports non-existent property instead

### ISSUE THREAD

**Issue by DidierLoiseau** [assertj/assertj#3563]
**Describe the bug**
If a getter throws an exception, it is not reported in the `AssertionError` from `hasFieldOrPropertyWithValue()`

* assertj core version: 3.26.3
* java version: 11
* test framework version: junit 5.10.2

**Test case reproducing the bug**

```java
import org.junit.jupiter.api.Test;

import static org.assertj.core.api.Assertions.assertThat;

class ThrowingGetterTest {
  @Test
  void test() {
    assertThat(new MyThrowingPojo())
        .hasFieldOrPropertyWithValue("property", "value");
  }

  static class MyThrowingPojo {
    public String getProperty() {
      throw new RuntimeException();
    }
  }
}
```
Output:
```
java.lang.AssertionError: 
Expecting
  AssertJTest$MyThrowingPojo@be64738
to have a property or a field named "property"
(static and synthetic fields are ignored)

	at AssertJTest.test(AssertJTest.java:9)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
```
(no trace of the `RuntimeException` – note that the test succeeds when `getProperty()` returns `"value"`)

The problem appears to be that the nice `IntrospectionError` built by `PropertyOrFieldSupport#getSimpleValue()` is discarded in `org.assertj.core.internal.Objects#assertHasFieldOrProperty()`.

In my specific case, the object under test was implemented by proxy which was throwing the exception, which makes it really confusing.

**Comment by scordio** [comment]
Thanks for reporting the issue and providing a small reproducer, @DidierLoiseau!

**Comment by etrandafir93** [comment]
Hello @scordio, @joel-costigliola,

I took a quick look over this one. The issue seems to be in [Introspection::getPropertyGetter](https://github.com/assertj/assertj/blob/4c2d0b1b9dfb6b00c76f55fd155a0a2c844285bc/assertj-core/src/main/java/org/assertj/core/util/introspection/Introspection.java#L72) where we catch all exceptions and wrap them within a _IntrospectionError_ which is later treated as if the field is not present.

If the getters throw an exception, this will be wrapped within an _IllegalAccessException or InvocationTargetException_. So, what we can do is check the exception cause and - if needed - rethrow it without wrapping it in our [IntrospectionError](https://github.com/assertj/assertj/blob/4c2d0b1b9dfb6b00c76f55fd155a0a2c844285bc/assertj-core/src/main/java/org/assertj/core/util/introspection/IntrospectionError.java#L21)

```java
try {
      getter.setAccessible(true);
      getter.invoke(target);
} catch (IllegalAccessException | InvocationTargetException t) {
 
      if (t.getCause() == null || isCausedBy(t, IllegalAccessException.class, InvocationTargetException.class)) { 
        
          //  same as now
          throw new IntrospectionError(propertyNotFoundErrorMessage("Unable to find property %s in %s", propertyName, target), t);
           
         //  otherwise throw a new type of runtime exception that will not be caught by our logic
} else { 
        throw new GetterInvocationError(format("When invoking the getter %s, encountered exception with message: %s", getter.toGenericString(), t.getCause().getMessage()), t);
}

static class GetterInvocationError extends RuntimeException { ... }
```

If this solution seems reasonable, I can raise a PR so you can take a better look into it :) 
Let me know what you think.

**Comment by joel-costigliola** [comment]
@etrandafir93 sounds like a good plan to me, please go on with a PR, thanks!

**Comment by etrandafir93** [comment]
@joel-costigliola - This will be even simpler than I initially thought. 
We only need to have a different catch block for `InvocationTargetException` and wrap it into a different runtime ex:

```diff
try {
      getter.setAccessible(true);
      getter.invoke(target);
+ } catch (InvocationTargetException t) {
+       //  if getter invocation fails, throw a different exception that's not caught by our logic
+       throw new GetterInvocationError(formatErrorMessage(
+                 "Encountered exception when invoking getter for %s in %s", propertyName, target), t.getCause());
+
} catch (Exception t) {
      throw new IntrospectionError(formatErrorMessage("Unable to find property %s in %s", propertyName, target), t);
}
```

With this change, we'll still use the field as a fallback verification - but only when the getter invocation fails for reflection-related reasons. _Basically, for all the other exceptions described [here](https://docs.oracle.com/javase/1.5.0/docs/api/java/lang/reflect/Method.html#invoke(java.lang.Object,%20java.lang.Object...)), apart from InvocationTargetException_

---

On the other hand, I've noticed tests suggesting the original intended behavior was to bypass **ANY** exception thrown by the getter and verify the field:

For eg: [FieldsOrPropertiesExtractor_extract_Test::should_fallback_to_field_if_exception_has_been_thrown_on_property_access](https://github.com/assertj/assertj/blob/4c2d0b1b9dfb6b00c76f55fd155a0a2c844285bc/assertj-core/src/test/java/org/assertj/core/groups/FieldsOrPropertiesExtractor_extract_Test.java#L107)

In other words, if the getter throws an error we'll check for the field. If the field is also missing (the case described in this issue), we'll return the non-existent property assertion error, swallowing the exception.

So I guess it's a matter of deciding what the correct behavior would be in this case.

**Comment by joel-costigliola** [comment]
I think we want to keep the behavior that if there is no property we should be checking the field, otherwise it's going to be a sneaky breaking change, but if the property exists, and we can't get its value then I think it is reasonable to throw a specific error saying so.

If I understand your comment correctly, it is what you have done, right ?

**Comment by etrandafir93** [comment]
As things stand, we initially try to invoke the getter: If there is any sort of exception (regardless if it's a reflection-related exception in the lib or if it's an exception in the tested code, in the getter itself), we'll try to verify the field as a fallback. 

As a result, a test with the following setup will pass:

```java
static class MyPojo {
  private String name = "abcd";
  private String address = "abcd";

  public String getName() {
    throw new RuntimeException("some exception");
  }
  public String getAddress() {
    return "something else";
  }
}

@Test
void test() {
  assertThat(new MyPojo()).hasFieldOrPropertyWithValue("name", "abcd"); // <-- this is passing, is it ok?
  assertThat(new MyPojo()).hasFieldOrPropertyWithValue("address", "abcd"); // <-- this is failing
}
```

As I understand, we want to keep maintaining this behavior, not to introduce a breaking change - correct? 

_If I understand your comment correctly, it is what you have done, right ?_
Not yet, but yes - we can go into this direction. Please confirm if I got it right: 
 - maintain current behavior 
 - if the field is also missing, rethrow the original exception produced by the getter

---

_PS: Another interesting case I found (and maybe a bit unexpected/counterintuitive) is with the package-protected getters which will always be ignored, but seems to be slightly off-topic here._

**Comment by joel-costigliola** [comment]
You got it right, let's leave the package protected case to another discussion so that we tackle one issue at a time
